// @nodearch/core/dist/components/enums.js
export var ComponentScope;
(function (ComponentScope) {
    ComponentScope["REQUEST"] = "Request";
    ComponentScope["SINGLETON"] = "Singleton";
    ComponentScope["TRANSIENT"] = "Transient";
})(ComponentScope || (ComponentScope = {}));
export var CoreDecorator;
(function (CoreDecorator) {
    CoreDecorator["COMPONENT"] = "@nodearch/core/decorators/component";
    CoreDecorator["COMMAND"] = "@nodearch/core/decorators/command";
    CoreDecorator["HOOK"] = "@nodearch/core/decorators/hook";
    CoreDecorator["CONFIG"] = "@nodearch/core/decorators/config";
    CoreDecorator["SERVICE"] = "@nodearch/core/decorators/service";
    CoreDecorator["REPOSITORY"] = "@nodearch/core/decorators/repository";
    CoreDecorator["INTERCEPTOR"] = "@nodearch/core/decorators/interceptor";
    CoreDecorator["CONTROLLER"] = "@nodearch/core/decorators/controller";
    CoreDecorator["USE"] = "@nodearch/core/decorators/use";
})(CoreDecorator || (CoreDecorator = {}));
export var DecoratorType;
(function (DecoratorType) {
    DecoratorType["CLASS"] = "class";
    DecoratorType["CLASS_METHOD"] = "classMethod";
    DecoratorType["METHOD"] = "method";
    DecoratorType["PARAMETER"] = "parameter";
})(DecoratorType || (DecoratorType = {}));
